src/jetstream_types.rs

```diff
@@ -553,23 +584,214 @@ pub struct NextRequest {
-    pub stream: String,
-    pub consumer: String,
-    // For pull based consumers, batch size for pull
-    pub pull: usize,
-    // For manual ack
-    pub mack: bool,
-    // For creating or updating.
-    pub cfg: ConsumerConfig,
+pub(crate) struct StreamNamesResponse {
+    pub streams: Vec<String>,
+}
+
+/// Options for subscription
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+pub struct SubscribeOptions {
+    // For consumer binding:
+    pub(crate) bind_only: bool,
+    pub(crate) stream_name: Option<String>,
+    pub(crate) consumer_name: Option<String>,
+
+    // For consumer configuration:
+    pub(crate) ordered: bool,
+    pub(crate) ack_policy: Option<AckPolicy>,
+    pub(crate) ack_wait: Option<i64>,
+    pub(crate) replay_policy: Option<ReplayPolicy>,
+    pub(crate) deliver_policy: Option<DeliverPolicy>,
+    pub(crate) deliver_subject: Option<String>,
+    pub(crate) description: Option<String>,
+    pub(crate) durable_name: Option<String>,
+    pub(crate) sample_frequency: Option<u8>,
+    pub(crate) idle_heartbeat: Option<i64>,
+    pub(crate) max_ack_pending: Option<i64>,
+    pub(crate) max_deliver: Option<i64>,
+    pub(crate) max_waiting: Option<i64>,
+    pub(crate) opt_start_seq: Option<u64>,
+    pub(crate) opt_start_time: Option<DateTime>,
+    pub(crate) flow_control: Option<bool>,
+    pub(crate) rate_limit: Option<u64>,
+    pub(crate) headers_only: Option<bool>,
+}
+
+impl SubscribeOptions {
+    /// Creates a new set of default subscription options
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Binds to an existing consumer from a stream without attempting to create one.
+    pub fn bind(stream_name: String, consumer_name: String) -> Self {
+        Self {
+            stream_name: Some(stream_name),
+            consumer_name: Some(consumer_name),
+            bind_only: true,
+            ..Default::default()
+        }
+    }
+
+    /// Creates an ordered fifo (first-in-first-out) subscription.
+    pub fn ordered() -> Self {
+        Self {
+            ordered: true,
+            ..Self::default()
+        }
+    }
+
+    /// Binds the consumer to a stream explicitly based on a name.
+    ///
+    /// When a stream name is not specified, the subject is used as a way to find the stream name.
+    /// This is done by making a request to the server to get list of stream names that have a filter the subject.
+    /// To avoid the stream lookup, provide the stream name with this function.
+    pub fn bind_stream(stream_name: String) -> Self {
+        Self {
+            stream_name: Some(stream_name),
+            ..Default::default()
+        }
+    }
+
+    /// Sets the description used for the created consumer.
+    pub fn description(mut self, description: String) -> Self {
+        self.description = Some(description);
+        self
+    }
+
+    /// Sets the durable name for the created consumer.
+    pub fn durable_name(mut self, consumer: String) -> Self {
```

Comment:
Should durable mutually exclusive along with bind and ordered as a top level option?